Crop model GDD

nemozen opened the pull request:
CropModel.growing_degree_days()

nemozen committed:
add growing_degree_days function to CropModel

nemozen committed:
in find_data_series(), yield all possible series  in order instead of just the highest ranked one

add_data_series() will still only use the first one, but this gives
more flexibility when using the lower level function find_data_series

nemozen commented:
branched off https://github.com/gro-intelligence/api-client/pull/125

nemozen committed:
Merge remote-tracking branch 'origin/development' into crop_model_gdd_impl

jeanzhaoyi commented on api/client/crop_model.py:
```diff
@@ -69,3 +70,51 @@ def compute_crop_weighted_series(self,
+        [(T_max + T_min)/2 - T_base], and the GDD over a longer time
+        interval is the sum of the GDD over all days in the interval.
+
+        The region can be any region of the Gro regions, from a point
```
What if only Tmean is available for a region (say using LST) while daily Tmin and Tmax were not available or incomplete (ghcn stations)?

nemozen replied:
Yes good question.  The idea is that we use find_data_series to get the best version of GDD, and that depends on the combination of search ranking and series availability.

If the region is specific and the daily max and min exists for that exact region,  e.g. a weather station  with GHCN data, then it does exactly what you would expect.   If not , then it depends on what is avialable.. If we don't have max and min temp but we do have the daily mean temperature, then  that should be the closest matching item and metric, and taking the mean of that should be ok. Or if we have more samples uniformly throughout the day  e.g. temperature every hour, then takign the average ofa all of them should be ok too. But note that *land surface* temperature is different than *air*  temperature so that would be wrong. 
Similarly, if the region is higher level, then it should give preference to weather stations in that district, rather than picking other lst at the district level.

Updataed the code to make the generalization clearer.

jeanzhaoyi replied:
Often there are some GHCN stations with available data. The issue is that GHCN station data tend to be highly sparse temporally. For example, there might be data for 2 years or a few days a year, while the desired time period is spreading over 10 years. I think there needs to be a judgement call on a minimum threshold of temporal coverage for a region required to use the data from a source. Like you said, air temperature is very different from land surface temperature, so we need to be consistent with choosing one temperature source for a given spatial extent (likely a list of districts or provinces) and time period.

nemozen replied:
Yup, I was planning on adding coverage threshold later. But I've added it to this PR.

nemozen committed:
update gdd function to be more general mean (equivalent to previous for tmin, tmax)

nemozen committed:
add temporal coverage threshold

nemozen committed:
Merge remote-tracking branch 'origin/development' into crop_model_gdd_impl

nemozen committed:
improve handling of empty df when insufficient data

jeanzhaoyi commented on api/client/crop_model.py:
```diff
@@ -69,3 +70,75 @@ def compute_crop_weighted_series(self,
+        min_temporal_coverage: float, optional
+
+        """
+        for tmax in self.find_data_series(
```
Since we are not specifying the source_id here, it could be LST or GHCN or some other future sources right?

nemozen replied:
Right. It will use  the source ranking for that time range and region.

jeanzhaoyi replied:
In that case, a user would bear the risk of having their daily mean temperature calculated from mixed sources. The risk being that certain regions or dates with more availability of one source have a more biased temperature than others, since LST tends to have higher Tmean than the GHCN Tmean. I think the current script is fine if a user can ignore such biases or uncertainties attributed to various data availability scenarios. Meanwhile, a user should be able to specify one particular source to use exclusively as well, like an argument `source_name` or `source_id` with default value Null, in which case the source ranking decides the source(s).

nemozen replied:
Right.. 

growing_degree_days()  uses find_data_series()  entity search and series availability  to  avoid the user having to know which specific series they need, which depends on the region and the date ranges. 

I've separated the lower level calculation into a separate function compute_gdd() so a user who already knows how to select specific series can do it directly.

nemozen committed:
expose lower level compute_gdd() function to separate finding data series from computation

this allows sophisticated users to control which series are used
instead of depending on find_data_series

nemozen committed:
CLEWS-18837:  more complete docstrings for CropModel

hongwyx commented on api/client/crop_model.py:
```diff
@@ -69,3 +119,83 @@ def compute_crop_weighted_series(self,
+            raise Exception("Insufficient temporal coverage for GDD, " + \
+                            "{} < {} data points available".format(
+                                tmean.value.size, coverage_threshold))
+        gdd_values = tmean.value - base_temperature
```
A few comments to ensure that the function works according to context:

The function currently allows for GDD to be unbounded; however, according to the source below, GDD shouldn't be negative, and should have an upper bound. At the very least I think we should not let it be negative, by modifying 

`gdd_values = tmean.value - base_temperature`

to 

`gdd_values = tmean.value.apply(lambda x: min(x - base_temperature, 0))`

and also require base_temperature to be non-negative (or within a certain range), otherwise raise exception.

When there is no temperature data for a particular day, it would be better to raise a warning so that the user knows that the GDD is not accumulating due to missing data, instead of ineligible temperature range.

Also, not sure if all temperature data series that we pull share the same unit (Fahrenheit/Celsius?) We should specify that in case the user is confused about what to use as the base temperature. (Do we have a unit converter function?)

(Additionally, we could also require start and end date to be within the same growing season.)

Attached is my script where I tested a few input combinations. I modified the import path/generator syntax (`next(...)` instead of `(...).next()` )to make it work with Python 3.


Sources:
http://agron-www.agron.iastate.edu/courses/Agron541/classes/541/lesson02b/2b.1.1.html

jeanzhaoyi replied:
If the temperature at a given day, say `tmean.value`, is lower than the `base_temperature`, the contribution to GDD on that day should be zero.
Here's how I calculated GDD: 
```
(( (temp % up_thresh) * (temp < up_thresh) + 
(temp > up_thresh) * up_thresh) - lower_thresh) * (temp >= lower_thresh)
```

hongwyx replied:
Modified to match your expression.

nemozen replied:
thanks!

- changed each gdd to >= 0 
- added support for upper_temperature_cap
- changed default min_temporal_coverage to 1.0 so by default we do not allow any missing days at all
- added warning when time range > 1 year

nemozen committed:
take positive part in gdd defn

nemozen committed:
more conservative, make default min temporal coverage 100%

with a > 0 gdd, treating missing values as 0 on average could mislead naive users

nemozen committed:
clean up docstring

nemozen committed:
add warning if GDD time range > 1 year

Most crop seasons are < 1 year. But we make it a warning rather than
throwing an exception, in case advanced users want to compute it over
multiple years and average for example, or deal with crop seasons that
are multiyear

nemozen committed:
add optional upper_temperature_cap

nemozen committed:
Merge remote-tracking branch 'origin/development' into crop_model_gdd_impl

jeanzhaoyi reviewed (approved):
Looks good to me =)

nemozen committed:
revert unrelated change from another branch

nemozen committed:
Merge remote-tracking branch 'origin/development' into crop_model_gdd_impl
